### PR TITLE
Close Redis Broker and PresenceManager on Node shutdown

### PR DESCRIPTION
**Pull request by FZambia**
This closes all the running goroutines of Redis Broker on Node Shutdown. This should also fix our Survey tests when those are launched with count > 1.

### CODE REVIEW

**Comment by FZambia** [comment]
Redigo Conn does not support concurrent Receive and Close so we catched a data race here. Actually the fix in dac8a601ee24789d3076993f514009080adbf32b is to Unsubcribe from all channels and thus unblock the ReceiveWithTimeout call. I am not really happy with it, possibly with `go-redis` we can just close the connection concurrently.

**Comment by j178** [comment]
With this patch, I was able to runthis benchmark successfully on my windows desktop (have to manually increase the redis `maxclients` configuration). But sadly it still doesn't work on my mac. I guess it's because my windows desktop has a more powerful spec and better performance. Anyway, I think this issue is fixed.

**Comment by FZambia** [comment]
> But sadly it still doesn't work on my mac.

I am also using Mac, probably not everything is still cleaned up in tests, I tried to monitor number of conns to Redis using:

```
echo 'client list' | redis-cli | wc -l
``` 

And it constantly grows, up to almost 10k conns. Will try to find out why this happens.

**Comment by FZambia** [comment]
61fdb9ec375a206a3c8d25d9ca5767aaeabe52d9 reduces number of conns during the test run, for me the number of conns to Redis does not exceed 2.2k during the test now. So I think with big enough `maxclients` and large `ulimit -n` tests should pass. Probably I'll look a bit more whether 2.2k is a reasonable number for this test or still something not cleaned up - since still looks to many even for 100 nodes participating in test.

**Comment by FZambia** [comment]
Uff, it's harder than expected, not sure whether it's possible to achieve graceful shutdown with Redigo in non-racy way and keep current engine semantics, but I'll keep trying.

**Comment by FZambia** [comment]
@j178 I believe should be fine now, conns properly closing and not growing at all during the test with the latest changes.

**Comment by j178** [comment]
@FZambia Wow, amazing work! It works for me now!

**Comment by FZambia** [comment]
Cool, thx for checking. I'd like to test it out a bit more - want to look closer at `conn.Unsubscribe()` behaviour. Wondering whether it's possible to avoid `conn.Unsubscribe` in `go-redis` case and simply close PUB/SUB connection concurrently with `Receive` - I think it's much more lightweight way to quit from PUB/SUB routine.

File: broker_redis.go [hunk 1]
```diff
@@ -932,59 +1016,67 @@ func (b *RedisBroker) runPubSubPing(s *RedisShard) {
...
-		conn := s.pool.Get()
-		err := conn.Send("PUBLISH", b.pingChannel, nil)
-		if err != nil {
-			b.node.Log(NewLogEntry(LogLevelError, "error publish ping to Redis channel", map[string]interface{}{"error": err.Error()}))
-			_ = conn.Close()
+		select {
+		case <-b.closeCh:
 			return
+		case <-pingTicker.C:
+			<-pingTicker.C
```
Comment (j178) [inline broker_redis.go:1023]: This line is duplicated I think.

Comment (FZambia) [reply broker_redis.go:1023]: Great catch - thank you!

**Pull request merged**